### PR TITLE
Apple TV support

### PR DESCRIPTION
**Pull request by dlowder-salesforce**
**Motivation**

Adding support for Apple TV.  See README-appletv.md for implementation details.

**Test plan (required)**

All existing tests pass in the new Apple TV projects.  There are additions to scripts/objc-test.sh to automatically run tests in both the iOS and Apple TV projects.

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @adamjernst and @nathanajah to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by bestander** [comment]
@dlowder-salesforce, that is crazy cool!

**Comment by bestander** [comment]
Did you discuss it with other core contributors?

The concerns that raise every time there are new form factors / OS added:
- do we need to add another .js suffix for packager to recognize
- who will support issues and features for this platform

Afaik we agreed with MS people to keep their [fork](https://github.com/ReactWindows/react-native-windows) separate for now because of the concerns above.

Once we finish reworks on the packager, we hope that plugging new platforms into JS pakcager will be easier, then it will all come to who owns the platform support.

**Comment by javache** [comment]
Thanks for this contribution, it's great to see that AppleTV can be supported this way. As has been mentioned before, we really do need to find a more scalable way to add more platforms. Every platform needing its own packager param, its own globals and its own js suffix is not something that scales.

If we decide to try to merge this in the core repo, could you please also put this up as multiple smaller pull requests that each represent a functional unit of work?

**Comment by douglowder** [comment]
@bestander @javache : 

- I have not discussed this outside Salesforce until now, as I had to wait for internal approvals and for Salesforce to sign the CLA before submitting this request.
- I understand your concerns about scalability.  The good news is that this isn't really adding an entire new platform -- it's really just a variant of iOS (it shares > 95% of its code in common with iOS) so it wouldn't require the same effort to maintain as an entirely new platform like MS.  I've already added automated testing for the new Apple TV projects.
- I've made the necessary changes to the existing packager already.  Do you have an estimate for when your packager rework will be pushed to master?
- I will have a look at my changes to see if there is a natural way to break them up into smaller pull requests.

**Comment by bestander** [comment]
> I've made the necessary changes to the existing packager already. Do you have an estimate for when your packager rework will be pushed to master?

Packager maintainers were really unhappy when I approved a hack to support ".windows.js" to the packager internals.
We need to come up with a more pluggable way, ideally every apps should be able to register how a suffix should be interpreted.

We are reworking the packager during the rest of 2016 and our main priority is performance and maintainability. 
We will have a separate discussion for what we will do with all the platforms in the [Core Contributors Group](https://www.facebook.com/groups/reactnativeoss/), no date yet, but feel free to start it.

cc @matryoshcow who works on packager full time now.

File: Examples/2048/2048TV.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -0,0 +1,429 @@
+// !$*UTF8*$!
+{
```
Comment (bestander) [inline Examples/2048/2048TV.xcodeproj/project.pbxproj]: These files are very large and will be hard to maintain

File: React/Views/RCTWebViewManager.m [hunk 2]
```diff
@@ -14,6 +14,8 @@
 #import "RCTWebView.h"
 #import "UIView+React.h"
 
+#if !TARGET_OS_TV
```
Comment (bestander) [inline React/Views/RCTWebViewManager.m]: These #if flags make ObjC code more complex and when we add new components people will forget to add them.

Comment (douglowder) [inline React/Views/RCTWebViewManager.m]: True, however these flags are only needed for certain specific components (e.g. ones that map to native iOS APIs that don't exist in tvOS).  Most components shouldn't need them.

File: Examples/2048/2048TV.xcodeproj/project.pbxproj (hunk 1 above)
Comment (douglowder) [inline Examples/2048/2048TV.xcodeproj/project.pbxproj]: They are identical to the corresponding iOS projects except for a few lines... maybe there's a way they could be generated from a single file, or combined.

**Comment by douglowder** [comment]
@bestander @matryoshcow Good news is this change doesn't require a new .js extension for the packager, it just uses the existing *.ios.js extensions.

File: Examples/2048/2048TV.xcodeproj/project.pbxproj (hunk 1 above)
Comment (bestander) [inline Examples/2048/2048TV.xcodeproj/project.pbxproj]: Yeah, the goal is to have proj files generated by Buck.
We'll be working on this in the following months

**Comment by dlowder-salesforce** [comment]
@javache @bestander In looking at how to break this work up into smaller pull requests, I came up with the following plan:

PR 1: Packager changes to add the new Apple TV flag for the iOS platform.
PR 2: Adding TARGET_OS_TV defines in iOS source code to ensure that code compiles in tvOS.
PR 3: Add tvOS project files for the example projects and for react-native-cli.
PR 4: Changes to RCTView, and RCTViewManager to add support for the tvOS focus engine, and adding the TV gesture recognizer class.
PR 5: Changes to Libraries/Components and Examples to add tvOS focus engine support in the JS layer.
PR 6: Add tvOS target to scripts/objc-test.sh.

What do you think?

**Comment by facebook-github-bot-5** [comment]
@dlowder-salesforce updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9577/files/4f7e2836636f63b99cf5a5d36357d8eb5c79795e..38fbba6a19ed4924925588359a799db2dbfa21ec)

**Comment by dlowder-salesforce** [comment]
> We will have a separate discussion for what we will do with all the platforms in the Core Contributors Group, no date yet, but feel free to start it.

Thanks for this suggestion @bestander ... my request to join this group is pending.  Once it's accepted, I'll start that discussion.

**Comment by bestander** [comment]
Hey @dlowder-salesforce, you are now added to the group.

**Comment by dlowder-salesforce** [comment]
Closing this pull request, and breaking this change up into smaller commits as outlined in the comment above.

**Comment by dlowder-salesforce** [comment]
@bestander @matryoshcow As suggested, I posted in React Native Core Contributors to start a discussion about packager rework for new platforms.... no response yet.  Also, I've submitted the first of my six pull requests: https://github.com/facebook/react-native/pull/9649 . Leaving the packager changes until near the end of my commits to allow time for this discussion.

**Pull request closed without merging**